pkg/agent/nplagent/k8s/handlers.go

```diff
@@ -0,0 +1,135 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8s
+
+import (
+	"fmt"
+
+	nplutils "github.com/vmware-tanzu/antrea/pkg/agent/nplagent/lib"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/klog"
+)
+
+func (c *Controller) addRuleForPod(pod *corev1.Pod) {
+	podIP, nodeIP := pod.Status.PodIP, pod.Status.HostIP
+	if podIP == "" || nodeIP == "" {
+		return
+	}
+	podContainers := pod.Spec.Containers
+
+	for _, container := range podContainers {
+		for _, cport := range container.Ports {
+			port := fmt.Sprint(cport.ContainerPort)
+			nodePort, ok := c.portTable.AddRule(podIP, int(cport.ContainerPort))
+			if !ok {
+				klog.Warningf("Failed to add rule for podIP: %s, port: %d", podIP, cport.ContainerPort)
+				continue
+			}
+			assignPodAnnotation(pod, port, nodeIP, fmt.Sprint(nodePort))
+		}
+	}
+}
+
+// HandleAddPod handles pod annotations in NPL for an added pod
+//func (c *Controller) HandleAddPod(pod *corev1.Pod) {
+func (c *Controller) HandleAddPod(obj interface{}) {
+	klog.Infof("Got add pod")
+	pod := obj.(*corev1.Pod).DeepCopy()
+	klog.Infof("Got add event for pod: %s/%s", pod.Namespace, pod.Name)
+	c.addRuleForPod(pod)
+	if pod.Annotations[NPLAnnotationStr] != "" {
+		c.updatePodAnnotation(pod)
+	}
+}
+
+// HandleDeletePod handles pod annotations for a deleted pod
+//func (c *Controller) HandleDeletePod(pod *corev1.Pod) {
+func (c *Controller) HandleDeletePod(obj interface{}) {
+	pod := obj.(*corev1.Pod).DeepCopy()
+	klog.Infof("Got delete event for pod: %s/%s", pod.Namespace, pod.Name)
+	podIP := pod.Status.PodIP
+	if podIP == "" {
+		return
+	}
+
+	for _, container := range pod.Spec.Containers {
+		for _, cport := range container.Ports {
+			c.portTable.DeleteRule(podIP, int(cport.ContainerPort))
+		}
+	}
+}
+
+// HandleUpdatePod handles pod annotations for a updated pod
+//func (c *Controller) HandleUpdatePod(old, newp *corev1.Pod) {
+func (c *Controller) HandleUpdatePod(oldObj, newObj interface{}) {
+	oldPod := oldObj.(*corev1.Pod).DeepCopy()
+	newPod := newObj.(*corev1.Pod).DeepCopy()
+
+	klog.Infof("Got update for pod: %s/%s", newPod.Namespace, newPod.Name)
+	podIP := newPod.Status.PodIP
+
+	if podIP == "" {
+		return
+	}
+
+	var newPodPorts []string
+	newPodContainers := newPod.Spec.Containers
+	for _, container := range newPodContainers {
+		for _, cport := range container.Ports {
+			port := fmt.Sprint(cport.ContainerPort)
+			newPodPorts = append(newPodPorts, port)
+			if !c.portTable.RuleExists(podIP, int(cport.ContainerPort)) {
+				c.addRuleForPod(newPod)
+			}
+		}
+	}
+
+	// oldPodPorts: [8080, 8081] newPodPorts: [8082, 8081] portsToRemove should have: [8080]
+	oldPodContainers := oldPod.Spec.Containers
+	oldPodIP := oldPod.Status.PodIP
+
+	if oldPodIP != "" {
+		for _, container := range oldPodContainers {
+			for _, cport := range container.Ports {
+				port := fmt.Sprint(cport.ContainerPort)
+
+				if !nplutils.HasElem(newPodPorts, port) {
```

Comment:
Do we need a generic slice lookup func for looking up a port string? Why not use a map for newPodPorts?